File: minesweeper/java-minesweeper/src/main/java/it/chicio/minesweeper/field/resolver/FieldResolverByIteratingThroughEachValue.java
package it.chicio.minesweeper.field.resolver;

import it.chicio.minesweeper.field.Field;

public class FieldResolverByIteratingThroughEachValue implements FieldResolver {
    public Field resolve(Field field) {
        Field resolvedField = new Field(field.numberOfRows(), field.numberOfColumn());
        for (int row = 0; row < field.numberOfRows(); row++) {
            for (int column = 0; column < field.numberOfColumn(); column++) {
                if (isABomb(field.get(row, column))) {
                    resolvedField.set(row, column, "*");
                } else {
                    resolvedField.set(row, column, String.valueOf(getNumberOfBombsInNeighbours(field, row, column)));
                }
            }
        }
        return resolvedField;
    }

    private boolean isABomb(String value) {
        return value.equals("*");
    }

    private int getNumberOfBombsInNeighbours(Field field, int row, int column) {
        int numberOfBombsNearCell = 0;
        for (String value : field.neighboursOf(row, column)) {
            if (isABomb(value)) {
                numberOfBombsNearCell++;
            }
        }
        return numberOfBombsNearCell;
    }
}

File: minesweeper/kotlin-mockk-minesweeper/src/main/kotlin/it/chicio/minesweeper/field/resolver/FieldResolverByIteratingThroughEachValue.kt
package it.chicio.minesweeper.field.resolver

import it.chicio.minesweeper.field.Field

class FieldResolverByIteratingThroughEachValue : FieldResolver {
    private val bombMarker = "*"

    override fun resolve(field: Field): Field {
        val resolvedField = Field(field.numberOfRows, field.numberOfColumn)
        (0 until field.numberOfRows).forEach { row ->
            (0 until field.numberOfColumn).forEach { column ->
                when {
                    isABomb(field[row, column]) -> resolvedField[row, column] = bombMarker
                    else -> resolvedField[row, column] = getNumberOfBombsInNeighbours(field, row, column).toString()
                }
            }
        }
        return resolvedField
    }

    private fun isABomb(value: String?): Boolean = value == bombMarker

    private fun getNumberOfBombsInNeighbours(field: Field, row: Int, column: Int): Int =
            field.neighboursOf(row, column).fold(0, { sumOfNeighbours, cell ->
                when {
                    isABomb(cell) -> sumOfNeighbours + 1
                    else -> sumOfNeighbours
                }
            })
}